### PR TITLE
Parameter defaults + lazy key creation

### PR DESCRIPTION
**Pull request by koosemose**
Implemented Parameter Defaults and now when a nonexistent key is accessed it is added to contents.

Additionally, all Parameters will have the contents dictionary, but only register as HasContents if there are entries in the Dictionary.

### CODE REVIEW

File: Assets/Scripts/Models/Parameter.cs [hunk 1]
```diff
@@ -21,11 +21,14 @@ public class Parameter {
...
     // If this Parameter contains other Parameters, contents will contain the actual parameters
     private Dictionary<string, Parameter> contents;
-
+    // Tracks if value has been explicitly set
+    private bool uninitializdeValue = true;
     public Parameter(string name, string value) 
     {
         this.name = name;
         this.value = value;
+        contents = new Dictionary<string, Parameter>();
```
Comment (mikejbrown) [inline Assets/Scripts/Models/Parameter.cs]: Why not `this.contents`? Just a little confused here.

File: Assets/Scripts/Models/Parameter.cs [hunk 2]
```diff
@@ -21,11 +21,14 @@ public class Parameter {
     private string value;
     // If this Parameter contains other Parameters, contents will contain the actual parameters
     private Dictionary<string, Parameter> contents;
-
+    // Tracks if value has been explicitly set
+    private bool uninitializdeValue = true;
```
Comment (mikejbrown) [inline Assets/Scripts/Models/Parameter.cs]: Spell fix `uninitializdeValue` -> `uninitializedValue`

**Comment by mikejbrown** [comment]
Thanks for the fixes. I'm just a little too tired to wrap my head around them at the moment :smile:. It occurs to me that this is the sort of data driven class that would be a perfect candidate for unit testing. Should eventually write up a set of test cases for this.

File: Assets/Scripts/Models/Parameter.cs (hunk 1 above)
Comment (koosemose) [inline Assets/Scripts/Models/Parameter.cs]: I tend to only use this when explicitly needed, such as here, where the name and value parameters have the same name as class fields, whereas contents isn't so no need to differentiate.

File: Assets/Scripts/Models/Parameter.cs (hunk 2 above)
Comment (koosemose) [inline Assets/Scripts/Models/Parameter.cs]: I'll fix this one if MonoDevelop ever stops rebelling.

File: Assets/Scripts/Models/Parameter.cs [hunk 3]
```diff
@@ -62,7 +71,8 @@ public Parameter(Parameter other)
             if (contents == null || contents.ContainsKey(key) == false)
             {
                 Debug.ULogWarningChannel("Parameter", "Trying to access non-existent key: '" + key + "'. Probably a Lua problem.");
-                return new Parameter(key); // return empty parameter for now
+                // Add a new blank key to contents, that will then be returned.
+                contents.Add(key, new Parameter(key));
```
Comment (mikejbrown) [inline Assets/Scripts/Models/Parameter.cs]: Is there a possibility contents could be null here? I added the check for it in line 71 but I guess it'll always be set now. So the check could be removed?

**Pull request merged**